File: chapter1/003/main.go
package main

import (
	"fmt"
	"unicode/utf8"
)

func main() {
	fmt.Println(urlEncode("Mr John Smith    ", 13))
}

func urlEncode(s string, sentenceSize uint) string {
	bytes := make([]byte, len(s))
	perTwenty := []byte("%20")
	i := 0 // Walks the original string
	j := 0 // Walks the new string
	for k := uint(0); k < sentenceSize; k++ {
		r, size := utf8.DecodeRuneInString(s[i:])
		if r == ' ' {
			bytes = append(bytes[:j], perTwenty...)
			j += len(perTwenty)
		} else {
			c := string(r)
			bytes = append(bytes[:j], c...)
			j += size
		}
		i += size
	}
	return string(bytes)
}

File: chapter1/003/main.py
def main():
    print(urlify("Mr John Smith    ", 13))

def urlify(string, realSize):
    string = list(string)

    i = len(string) - 1
    for j in range(realSize-1, -1, -1):
        if string[j] == " ":
            string[i-2:i+1] = ["%", "2", "0"]
            i -= 3
        else:
            string[i] = string[j]
            i -= 1
    
    return "".join(string)

main()